python/tskit/tables.py

```diff
@@ -3020,6 +3026,99 @@ def union(
                 record=json.dumps(provenance.get_provenance_dict(parameters))
             )
 
+    def time_join(self, recent, time, continue_nodes=None):
+        ancient_ts = self.tree_sequence()
+        recent_ts = recent.tree_sequence()
+        if continue_nodes is None:
+            continue_nodes = ancient_ts.samples()
+
+        new_tables = recent_ts.tables
+        new_nodes = np.empty(0, dtype=np.uint32)
+        for t in recent_ts.trees():
+            for root_id in t.roots:
+                root = recent_ts.node(root_id)
+                if root.time == time:
+                    new_nodes = np.append(new_nodes, root_id)
+                elif root.time < time:
+                    n = new_tables.nodes.add_row(time=time)
+                    new_tables.edges.add_row(
+                        left=t.interval.left,
+                        right=t.interval.right,
+                        parent=n,
+                        child=root_id,
+                    )
+                    new_nodes = np.append(new_nodes, n)
+                else:
+                    raise RuntimeError(
+                        f"Root found older than join time (age = {root.time})"
+                    )
+        recent_ts = new_tables.tree_sequence()
+
+        if len(new_nodes) > len(continue_nodes):
+            raise RuntimeError(
+                f"""More lineages present at the longest-ago end of `recent_ts` \
+                than there are nodes in `continue_nodes`. \
+                Setting sample_size={len(continue_nodes)} or leaving it to default \
+                will ensure compliance."""
+            )
+
+        node_map = np.repeat(tskit.NULL, recent_ts.num_nodes)
+
+        node_map[new_nodes] = np.random.choice(
+            continue_nodes, len(new_nodes), replace=False
+        )
+
+        tables = ancient_ts.tables
+        tables.set_samples(continue_nodes, samples=False)
+        tables.union(
+            recent_ts.tables,
+            node_map,
+            add_populations=False,
+            check_shared_equality=False,
+            shift_time=time,
+        )
+
+        return tables.tree_sequence()
```

Comment:
If this is a method of ``TableCollection`` it should update the current table collection *in place*, not return a tree sequence. So, the call to ``union`` should be ``self.union()``

Reply:
Should it also return the tables after the call to `self.union()` or simply update in place? (should both `ancient_tables.time_join(recent_ts)` and `joined_tables = ancient_tables.time_join(recent_ts)` work?)

Reply:
Update in place - this is the convention we've adopted for these table collection operations.

We can add a ``TreeSequence.time_join`` method that would return a new tree sequence (but we don't need to do that here).